Adds Batch Calling

zkagin opened the pull request:
Depends on #35 
Fixes #17

Test Plan: All local tests pass, manually verified that a run of main.py produces seemingly correct data into the DB.

This diff adds batch calling. It's a fairly substantial rewrite of the API, but the component changes are:
 - Pushing params into the `request_data()` function so that they are more stateless and can be queued in a batch request.
 - Pushing sql and config params into the initialize function so it is easily accessible everywhere and doesn't need to be passed as a param so much.
 - Accumulating all of the batch requests and sending them off all at once to a batch object.
 - Grouping those batches and adding params for batch size (by query) and page size, with defaults based on our own optimization.
 - Handling those batches as they come in through the callback.
 - Add a fake batch object for testing purposes.
 - Cleaning up logging so that it is not as noisy when it is not in debug mode.

A few notes on implementation:
 - The callback is a nested method per the patterns seen in Google documentation.
 - The `batch.execute()` hangs until all of the callbacks have completed, so that gives the response time to add more calls to the queue in cases where there is another page. This is why the while loop works.

@dchess This should be tested on production before we consider merging so that we understand whether it solves some of our performance issues. Once this is ready, please let me know the before/after on a full run.

dchess commented:
@zkagin I can filter the course ids and test in production but we currently have over 1200 active course ids so we'll need to handle batches larger than 1000.

zkagin commented:
@dchess Oh wow, that is a lot more than I expected. Just updated to batch by 1000 at a time. Let me know if that works for you!

zkagin commented:
@dchess Take another look through the code and try locally testing. I added some more params for batch size and page size, cleaned up logging and put more stuff in debug mode, and did some general params cleanup.

Also, note that this still depends on the testing PR from before, so we should make sure to merge that one first.

dchess commented on README.md:
```diff
@@ -49,6 +49,10 @@ PULL_GUARDIANS=
 PULL_GUARDIAN_INVITES=
 
+# Batch parameters. Can be configured and changed to optimize performance.
+BATCH_SIZE=The number of dates or courses to batch at a time. MAX: 1000.
```
@zkagin The batch size necessary to adequately run the submissions is much smaller than what the other endpoints can handle. I'm wondering if rather than a global constant, these should be attributes of the class?

zkagin replied:
@dchess Fixed to make these by class.

zkagin commented:
@dchess Take another look and see if this works!

dchess commented on api.py:
```diff
@@ -15,17 +17,18 @@ class EndPoint:
-        self.date = None
         self.request_key = None
         self.table_name = f"GoogleClassroom_{self.classname()}"
+        self.batch_size = config.DEFAULT_BATCH_SIZE
```
@zkagin if this is always overridden by the child class config setting, why set this from the config at all?

zkagin replied:
@dchess The way I was thinking about it was that it would keep things functional if new endpoints are added. But after consideration, I think you're right that we can remove it here and it'll error for anyone adding a new endpoint until they set their own batch size.

I'll go ahead and make that change now.

dchess commented on api.py:
```diff
@@ -304,87 +334,104 @@ def __init__(self, service, school_year_start):
-                self.table_name, con=sql.engine, schema=sql.schema
+                self.table_name, con=self.sql.engine, schema=self.sql.schema
             )
             return courses.id.unique()
```
@zkagin Can you update line 344 to include this temp fix: 
```python
return sorted(courses.id.unique())
```
This will help address #40 until we can get a proper solution in place and make it possible to deploy this PR to prod after merging.

zkagin replied:
@dchess Done.
